### PR TITLE
Integrated public api for a new sample results page

### PR DESCRIPTION
**Pull request by dhakim87**
Added results page from source template.  

Screen Shot 2020-07-23 at 3 38 31 PM

### CODE REVIEW

File: microsetta_interface/implementation.py [hunk 1]
```diff
@@ -949,6 +951,30 @@ def post_update_sample(*, account_id=None, source_id=None, sample_id=None):
...
 
 
+@prerequisite([SOURCE_PREREQS_MET])
+def get_sample_results(*, account_id=None, source_id=None, sample_id=None):
+    has_error, source_output, _ = ApiRequest.get(
+        '/accounts/%s/sources/%s' %
+        (account_id, source_id)
+    )
+    if has_error:
+        return source_output
```
Comment (wasade) [inline microsetta_interface/implementation.py:961]: This was previously unsafe, see #10

File: microsetta_interface/implementation.py [hunk 2]
```diff
@@ -949,6 +951,30 @@ def post_update_sample(*, account_id=None, source_id=None, sample_id=None):
...
+        (account_id, source_id, sample_id))
+    if has_error:
+        return sample_output
+
+    return _render_with_defaults('sample_results.jinja2',
+                                 account_id=account_id,
+                                 source_id=source_id,
+                                 sample=sample_output,
+                                 source_name=source_output['source_name'],
+                                 taxonomy=SERVER_CONFIG["taxonomy_resource"],
```
Comment (wasade) [inline microsetta_interface/implementation.py:974]: I'm not sure I understand why this is a server configuration detail. What happens when shotgun data are available?

File: microsetta_interface/server_config.json [hunk 3]
```diff
@@ -1,6 +1,9 @@
 {
   "endpoint": "http://localhost:8083",
   "private_api_endpoint": "http://localhost:8082/api",
+  "public_api_endpoint": "http://localhost:8084/results-api",
+  "taxonomy_resource": "greengenes",
```
Comment (wasade) [inline microsetta_interface/server_config.json:5]: Aren't the availalbe resources discoverable from the public api?

File: microsetta_interface/static/css/4_column_flex.css [hunk 4]
```diff
@@ -0,0 +1,29 @@
+  * {
```
Comment (wasade) [inline microsetta_interface/static/css/4_column_flex.css:1]: It looks like the contents of this file are copied from somewhere else on the web. Please make sure to add a comment for attribution, and if there is a license to verify it is compatible with BSD-3.

File: microsetta_interface/templates/sample_results.jinja2 [hunk 5]
```diff
@@ -0,0 +1,104 @@
+{% extends "sitebase.jinja2" %}
+{% set page_title = "Sample Results" %}
+{% set show_breadcrumbs = True %}
+{% set show_logout = True %}
+
+{% block head %}
+  <link rel="stylesheet" type="text/css" href="https://cdn.datatables.net/1.10.21/css/jquery.dataTables.css" />
```
Comment (wasade) [inline microsetta_interface/templates/sample_results.jinja2:7]: should all these sources come from `static/`?

File: microsetta_interface/templates/sample_results.jinja2 [hunk 6]
```diff
@@ -0,0 +1,104 @@
...
+    let taxonomy = "{{taxonomy}}";
+    let metric = "{{alpha_metric}}";
+    let sampleId = "{{sample.sample_id}}";
+    var opt = {"renderer": "canvas", "actions": true};  /* Options for the Vega embedding */
+
+    function createDataTable() {
+        return function() {
+            $('#taxonomyTable').DataTable({
+                ajax: {
+                    url: "{{public_endpoint}}/taxonomy/present/single/" + taxonomy + "/" + sampleId
```
Comment (wasade) [inline microsetta_interface/templates/sample_results.jinja2:26]: Barcodes in Microsetta do not contain the Qiita study prefix. I'm not seeing anywhere that the `10317.` is being prepended?

File: microsetta_interface/templates/sample_results.jinja2 [hunk 7]
```diff
@@ -0,0 +1,104 @@
...
+</div>
+<div class="row">
+  <div class="column" style="text-align: center">
+    <h3> What is in your sample? </h3>
+    <div style="padding:10px">
+      <table id="taxonomyTable" class="display" style="width:100%">
+        <thead>
+        <tr>
+          <th>Relative Abundance</th>
+          <th>Family</th>
```
Comment (wasade) [inline microsetta_interface/templates/sample_results.jinja2:91]: No domain, phylum, class, order?

File: microsetta_interface/templates/source.jinja2 [hunk 8]
```diff
@@ -133,6 +136,9 @@
                 <div class="col-sm">
                     {{sample.sample_site |e}}
                 </div>
+                <div class="col-sm">
+                    <a class="btn btn-outline-info" href="/accounts/{{account_id}}/sources/{{source_id}}/samples/{{sample.sample_id}}/results">View</a>
```
Comment (wasade) [inline microsetta_interface/templates/source.jinja2:140]: Is there a check to make sure there are results for the sample?

File: microsetta_interface/static/css/4_column_flex.css (hunk 4 above)
Comment (gwarmstrong) [inline microsetta_interface/static/css/4_column_flex.css:1]:
```suggestion
/* https://www.w3schools.com/css/tryit.asp?filename=trycss_mediaqueries_flex */
  * {
```

File: microsetta_interface/templates/sample_results.jinja2 (hunk 7 above)
Comment (gwarmstrong) [reply microsetta_interface/templates/sample_results.jinja2:91]: That info comes in from the API but left it out for brevity. Can include if you like.

Comment (wasade) [reply microsetta_interface/templates/sample_results.jinja2:91]: Yes, thanks. Participants will be asking about the other taxonomic levels so best to include domain -> genus

File: microsetta_interface/templates/source.jinja2 (hunk 8 above)
Comment (gwarmstrong) [reply microsetta_interface/templates/source.jinja2:140]: Seemed to have overlooked exposing and endpoint for checking if sample results exist in `microsetta_public_api`. Can wrap it in with a new feature. Something like `/results-api/diversity/alpha/exists?sample_id=some_id`

Comment (gwarmstrong) [reply microsetta_interface/templates/source.jinja2:140]: And likely a similar feature for taxonomy

File: microsetta_interface/templates/sample_results.jinja2 (hunk 6 above)
Comment (gwarmstrong) [reply microsetta_interface/templates/sample_results.jinja2:26]: Yeah... Also we did not pass the correct variable to `sampleId`. Rought draft 😅

File: microsetta_interface/templates/sample_results.jinja2 (hunk 5 above)
Comment (gwarmstrong) [reply microsetta_interface/templates/sample_results.jinja2:7]: What are the pros/cons of that vs. using a CDN?

File: microsetta_interface/templates/source.jinja2 (hunk 8 above)
Comment (wasade) [reply microsetta_interface/templates/source.jinja2:140]: okay. What happens if a result has an alpha diversity value but not a taxonomy value? (shouldn't happen, but technically a possible outcome)

File: microsetta_interface/templates/sample_results.jinja2 (hunk 5 above)
Comment (wasade) [reply microsetta_interface/templates/sample_results.jinja2:7]: Note sure TBH. @dhakim87?

File: microsetta_interface/server_config.json (hunk 3 above)
Comment (gwarmstrong) [reply microsetta_interface/server_config.json:5]: Sure.
If we wanted to use that the options I see would be something like:
1) Choose one of the available taxonomy resources. E.g., just grab the first one in the list, or lexicographical minimum of them. Caveat: Does not guarantee we will use the specific resource we would like, if there is more than one. Seems a little weird.
2) Give the user a select picker to choose from. But then this has the caveats:
    a) they either see whatever variable names you gave the resources in the microsetta-public-api config
    b) OR we have to manually remap the variable names you gave them to things that appear to make more sense to the user
    c) This option doesn't really make sense if there is only one taxonomy resource corresponding to the given sample
3) You manually configure the taxonomy resource that the users see on their 16S samples, as above. Caveat: is not as flexible in the long run.

Comment (wasade) [reply microsetta_interface/server_config.json:5]: Okay thanks. Let's continue fwd with (3) and revise as the needs of the page grow

File: microsetta_interface/templates/sample_results.jinja2 (hunk 6 above)
Comment (gwarmstrong) [reply microsetta_interface/templates/sample_results.jinja2]: Updated in https://github.com/biocore/microsetta-interface/pull/11/commits/fd7d375a6ac31e7cf08bace8f770b65c53089b37

File: microsetta_interface/templates/sample_results.jinja2 (hunk 7 above)
Comment (gwarmstrong) [reply microsetta_interface/templates/sample_results.jinja2:96]: Updated in https://github.com/biocore/microsetta-interface/pull/11/commits/fd7d375a6ac31e7cf08bace8f770b65c53089b37

File: microsetta_interface/templates/source.jinja2 (hunk 8 above)
Comment (gwarmstrong) [reply microsetta_interface/templates/source.jinja2]: https://github.com/biocore/microsetta-interface/pull/11/commits/f00f8c492500290f2a0afdc5b6734cc39bf37b63
Only populates the button if the data have existing alpha diversity and taxonomy
Screen Shot 2020-07-27 at 9 52 32 PM

File: microsetta_interface/templates/sample_results.jinja2 (hunk 5 above)
Comment (gwarmstrong) [reply microsetta_interface/templates/sample_results.jinja2]: Updated in https://github.com/biocore/microsetta-interface/pull/11/commits/07ad4628e717c96ae820900a0f2783a6ad09f844

**Pull request merged**